apps/opencs/view/render/terraintexturemode.cpp

```diff
@@ -0,0 +1,328 @@
+    std::size_t spacelocation = mCellId.find(space);
+    std::string slicedX = mCellId.substr (hashlocation+1, spacelocation-hashlocation);
+    std::string slicedY = mCellId.substr (spacelocation+1);
+    CSMWorld::CellCoordinates mCoordinates(stoi(slicedX), stoi(slicedY));
+
+    CSMDoc::Document& document = getWorldspaceWidget().getDocument();
+    CSMWorld::IdTable& landTable = dynamic_cast<CSMWorld::IdTable&> (
+        *document.getData().getTableModel (CSMWorld::UniversalId::Type_Land));
+    CSMWorld::IdTable& ltexTable = dynamic_cast<CSMWorld::IdTable&> (
+        *document.getData().getTableModel (CSMWorld::UniversalId::Type_LandTextures));
+
+    int textureColumn = landTable.findColumnIndex(CSMWorld::Columns::ColumnId_LandTexturesIndex);
+    CSMWorld::LandTexturesColumn::DataType mPointer = landTable.data(landTable.getModelIndex(mCellId, textureColumn)).value<CSMWorld::LandTexturesColumn::DataType>();
+    CSMWorld::LandTexturesColumn::DataType mNew(mPointer);
+
+    int xInCell ((hit.worldPos.x() - (stoi(slicedX)* cellSize)) * landTextureSize / cellSize);
+    int yInCell ((hit.worldPos.y() - (stoi(slicedY)* cellSize)) * landTextureSize / cellSize);
+
+    hashlocation = mBrushTexture.find(hash);
+    std::string mBrushTextureInt = mBrushTexture.substr (hashlocation+1);
+    int brushInt = stoi(mBrushTexture.substr (hashlocation+1))+1; // All indices are offset by +1
+
+    if (mBrushShape == 0) mNew[yInCell*landTextureSize+xInCell] = brushInt;
+    if (mBrushShape == 1)
+    {
+        for(int i=-mBrushSize/2;i<mBrushSize/2;i++)
+        {
+            for(int j=-mBrushSize/2;j<mBrushSize/2;j++)
+            {
+                if (xInCell+i >= 0 && yInCell+j >= 0 && xInCell+i <= landTextureSize  - 1 && yInCell+j <= landTextureSize - 1)
+                    mNew[(yInCell+j)*landTextureSize+(xInCell+i)] = brushInt;
+            }
+        }
+    }
+    float distance = 0;
+    if (mBrushShape == 2)
+    {
+        float rf = mBrushSize/2;
+        int r = (mBrushSize/2)+1;
+        for(int i = -r; i < r; i++)
+        {
+            for(int j = -r; j < r; j++)
+            {
+                distance = std::round(sqrt(pow((xInCell+i)-xInCell, 2) + pow((yInCell+j)-yInCell, 2)));
+                if (xInCell+i >= 0 && yInCell+j >= 0 && xInCell+i <= landTextureSize - 1 && yInCell+j <= landTextureSize  - 1 && distance <= rf)
+                    mNew[(yInCell+j)*landTextureSize+(xInCell+i)] = brushInt;
+            }
+        }
+    }
+    if (mBrushShape == 3)
+    {
+      // Not implemented
+    }
+
+    QVariant changedLand;
+    changedLand.setValue(mNew);
+
+    CSMWorld::CommandMacro macro (document.getUndoStack(), "Edit texture records");
+    QModelIndex index(landTable.getModelIndex (mCellId, landTable.findColumnIndex (CSMWorld::Columns::ColumnId_LandTexturesIndex)));
+
+    CSMWorld::TouchLandCommand* ltexTouch = new CSMWorld::TouchLandCommand(landTable, ltexTable, mCellId);
```

Comment:
Please don't do this. Push the Command directly to the macro. With your code, if the second constructor throws an exception, you have a resources leak.

Reply:
I did fix this already, but I'd be interested on deeper explanation on what happens with exception, and why it's a memory leak. TouchLandCommand is based on Qt's QUndoCommand and I've understood it wouldn't need a delete anyway. I'd guess it has something to do with new not matched with delete anyway, but how come it isn't a leak if "new CSMWorld::TouchLandCommand(landTable, ltexTable, mCellId)" is directly pushed to macro?